sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/nullExpressions.scala

```diff
@@ -354,12 +354,14 @@ case class IsNotNull(child: Expression) extends UnaryExpression with Predicate {
 
   override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
     val eval = child.genCode(ctx)
-    val value = eval.isNull match {
-      case TrueLiteral => FalseLiteral
-      case FalseLiteral => TrueLiteral
-      case v => JavaCode.isNullExpression(s"!$v")
+    val (value, newCode) = eval.isNull match {
+      case TrueLiteral => (FalseLiteral, EmptyBlock)
+      case FalseLiteral => (TrueLiteral, EmptyBlock)
+      case v =>
+        val value = ctx.freshName("value")
+        (JavaCode.variable(value, BooleanType), code"boolean $value = !$v;")
```

Comment:
This fix is related to https://github.com/apache/spark/pull/20965/files#diff-2eb948516b5beaeb746aadac27fbd5b4R287

Reply:
A simple query to reproduce this issue is as follows (without both changes);
sql("""
SELECT
  sum(CASE WHEN c0 IS NULL AND c1 IS NOT NULL THEN 1 ELSE 0 END) a,
  sum(CASE WHEN c0 IS NOT NULL AND c1 IS NOT NULL THEN 1 ELSE 0 END) b 
FROM
  VALUES ((null, null)) t(c0, c1)
""").show
```